pymc3/step_methods/hmc/integration.py

```diff
@@ -99,12 +103,15 @@ def _step(self, epsilon, state):
         # q_new = q + epsilon * v_new
         axpy(v_new, q_new, a=epsilon)
 
-        logp = self._logp_dlogp_func(q_new, q_new_grad)
+        p_new = RaveledVars(p_new, state.p.point_map_info)
+        q_new = RaveledVars(q_new, state.q.point_map_info)
+
+        logp = self._logp_dlogp_func(q_new, grad_out=q_new_grad)
 
         # p_new = p_new + dt * q_new_grad
-        axpy(q_new_grad, p_new, a=dt)
+        axpy(q_new_grad, p_new.data, a=dt)
 
-        kinetic = pot.velocity_energy(p_new, v_new)
+        kinetic = pot.velocity_energy(p_new.data, v_new)
         energy = kinetic - logp
 
         return State(q_new, p_new, v_new, q_new_grad, energy, logp)
```

Comment:
`q_new` and `p_new` might need to be packed back into `RaveledData` here, right?

Reply:
They should be `RaveledData` instances on that line (i.e. 117).

Note: I originally used an `ndarray` subclass to track `point_map_info`, instead of this `RaveledData` type.  It was more convenient to use, because these lines didn't need to change, but it would also hold onto the mapping information for longer than desired, and that didn't seem good.  I was able to make it return normal `ndarray`s whenever an operation changed its shape or dtype, but that wouldn't cover _everything_ so I scrapped it.